### PR TITLE
Bytearray

### PR DESCRIPTION
**Pull request by willmcgugan**
**What this PR solves**

- Uses bytearray and memoryview objects to reduce the amount of copying, and allowing masking to be done in-place

**Review**

- [X] Ready for review

### CODE REVIEW

File: lomond/mask.py [hunk 1]
```diff
@@ -10,77 +10,24 @@
...
+else:
+    _XOR_TABLE = [bytes(a ^ b for a in range(256)) for b in range(256)]
 
 
-if XorMaskerSimple is not None:
-    # Fast C version (works with Py2 and Py3)
-    def mask_payload(masking_key, data):
-        return XorMaskerSimple(masking_key).process(data)
+def mask_payload(masking_key, data):
+    """XOR mask bytes.
```
Comment (willmcgugan) [inline lomond/mask.py]: Apologies to @toudi for deleting his lengthy explanation, but it was hard to maintain.

File: tests/test_session.py [hunk 2]
```diff
@@ -288,209 +267,11 @@ def test_check_ping_timeout(session, mocker):
     assert session._check_ping_timeout(10, 11)
 
 
-@mocketize
```
Comment (willmcgugan) [inline tests/test_session.py]: mocket didn't support `recv_into`. Fortunately the integration tests with a tornado server have the same coverage.

File: lomond/mask.py [hunk 3]
```diff
@@ -10,77 +10,24 @@
...
-        # 0 1 2 3 4 5 6 7 8    ( byte no. )
-        # R O O O R O O O R
-        # then, for the second byte of the mask, we do:
-        # 0 1 2 3 4 5 6 7 8    ( byte no. )
-        # O R O O O R O O O
-        # please note, that even though byte 0 is marked as 'O', it has already
-        # been replaced in the previous step
-        return bytes(data_bytes)
-        # and voila!
+    """
```
Comment (willmcgugan) [inline lomond/mask.py]: Because we are working with mutable buffers, this can be done in-place. Saves two memcpys.

File: lomond/parser.py [hunk 4]
```diff
@@ -136,6 +135,7 @@ def _check_length(pos):
                 ParseError('unexpected eof of file')
             )
 
+        _buffer = self._buffer
```
Comment (toudi) [inline lomond/parser.py]: this is to avoid typing self. or does this does something under the hood?

Comment (willmcgugan) [reply lomond/parser.py]: It's a micro-optimization. `self._buffer` requires a dict lookup, but a local is faster.

**Pull request merged**